auth/clientcredentials/clientcredentials.go

```diff
@@ -0,0 +1,229 @@
+		}
+		return next(WithClientInfo(ctx, info), req)
+	}
+}
+
+func (i *Interceptor) WrapStreamingClient(next connect.StreamingClientFunc) connect.StreamingClientFunc {
+	return func(ctx context.Context, spec connect.Spec) connect.StreamingClientConn {
+		return next(ctx, spec) // no-op for clients
+	}
+}
+
+func (i *Interceptor) WrapStreamingHandler(next connect.StreamingHandlerFunc) connect.StreamingHandlerFunc {
+	return func(ctx context.Context, conn connect.StreamingHandlerConn) error {
+		if conn.Spec().IsClient {
+			return next(ctx, conn) // no-op for clients
+		}
+		requiredScopes, err := extractSchemaRequiredScopes(conn.Spec(), i.extension)
+		if err != nil {
+			return internalError(ctx, i.logger, err, "internal schema error") // invalid schema is internal error
+		}
+		info, err := i.requireScope(ctx, conn.RequestHeader(), requiredScopes)
+		if err != nil {
+			return err
+		}
+		return next(WithClientInfo(ctx, info), conn)
+	}
+}
+
+// RequireScope ensures the request context has a valid SAMS M2M token
+// with requiredScope. It returns a ConnectRPC status error suitable to be
+// returned directly from a ConnectRPC implementation.
+func (i *Interceptor) requireScope(ctx context.Context, headers http.Header, requiredScopes scopes.Scopes) (_ *ClientInfo, err error) {
+	var span trace.Span
+	ctx, span = tracer.Start(ctx, "clientcredentials.requireScope")
+	defer func() {
+		if err != nil {
+			span.RecordError(err)
+			span.SetStatus(otelcodes.Error, "check failed")
+		}
+		span.End()
+	}()
+
+	token, err := extractBearerContents(headers)
+	if err != nil {
+		return nil, connect.NewError(connect.CodeUnauthenticated,
+			errors.Wrap(err, "invalid authorization header"))
+	}
+
+	result, err := i.introspector.IntrospectToken(ctx, token)
+	if err != nil {
+		return nil, internalError(ctx, i.logger, err, "unable to validate token")
+	}
+	span.SetAttributes(attribute.String("client_id", result.ClientID))
+	info := &ClientInfo{ClientID: result.ClientID, TokenExpiresAt: result.ExpiresAt, TokenScopes: result.Scopes}
+	logger := i.logger.With(info.LogFields()...)
+
+	// Active encapsulates whether the token is active, including expiration.
+	if !result.Active {
+		// Record detailed error in span, and return an opaque one
+		span.SetAttributes(attribute.String("full_error", "inactive token"))
+		return info, connect.NewError(connect.CodePermissionDenied, errors.New("permission denied"))
+	}
+
+	// Check for our required scope.
+	for _, required := range requiredScopes {
+		if !result.Scopes.Match(required) {
+			// Record detailed error in span and logs
+			err = errors.Newf("got scopes %+v, required: %+v", result.Scopes, requiredScopes)
+			span.SetAttributes(attribute.String("full_error", err.Error()))
+			logger.Error("attempt to authenticate using SAMS token without required scope",
+				log.Error(err))
+			// Return an opaque error
+			return info, connect.NewError(connect.CodePermissionDenied, errors.New("insufficient scope"))
```

Comment:
I think:
1. Logging service-side errors here is non-actionable, after all this is client's problem to have insufficient scope.
2. The error returned to the client does not need to be opaque (no security concerns), and being explicit greatly help the client debug scope issues.

Reply:
Removed the log, and made the error non-opaque 👍